port traversability nodes summarily

Schmluk opened the pull request:
- Summarily ports the traversability places to develop (since they may be needed by other projects)
- Build and tests pass
- This mostly adds functionality and can not really interfere with existing bits I think 🙂

nathanhhughes commented on include/spark_dsg/traversability_boundary.h:
```diff
@@ -0,0 +1,411 @@
+
+namespace spark_dsg {
+
+/**
+ * @brief Check whether a state is considered traversable, i.e., it is TRAVERSABLE or
+ * TRAVERSED. If optimistic is true, UNKNOWN is also considered traversable.
+ */
+bool isTraversable(TraversabilityState state, bool optimistic = false);
+
+/** Check whether a set of states is considered traversable, i.e., all states are
+ * TRAVERSABLE or TRAVERSED. If optimistic is true, UNKNOWN is also considered
+ * traversable.
+ */
+bool isTraversable(const TraversabilityStates& states, bool optimistic = false);
+
+/**
+ * @brief Check whether a state is considered intraversable, i.e., it is INTRAVERSABLE.
+ * If optimistic is false, UNKNOWN is also considered intraversable.
+ */
+bool isIntraversable(TraversabilityState state, bool optimistic = false);
+
+/** Check whether a set of states is considered intraversable, i.e., at least one state
+ * is INTRAVERSABLE. If optimistic is false, UNKNOWN is also considered intraversable.
+ */
+bool isIntraversable(const TraversabilityStates& states, bool optimistic = false);
```
I sorta see the point for having both `isTraversable` and `isIntraversable`, but think `isIntraversable` does actually end up working out to be `!isTraversable`? Also the difference between State and States is subtle, I would do `areAllTraversable` instead of the overload

nathanhhughes commented on src/traversability_boundary.cpp:
```diff
@@ -0,0 +1,592 @@
+
+bool isIntraversable(TraversabilityState state, bool optimistic) {
+  return state == TraversabilityState::INTRAVERSABLE ||
+         (optimistic && state == TraversabilityState::UNKNOWN);
```
Shouldn't this clause be the inverse (i.e., `!optimistic && state == TraversabilityState::UNKNOWN`)?

nathanhhughes replied:
Possible I misunderstood the docstring though

nathanhhughes commented on src/traversability_boundary.cpp:
```diff
@@ -0,0 +1,592 @@
+    return true;
+  }
+
+  // TODO(lschmid): Implement.
+  return false;
```
Preference for no dead code if possible (I'd comment out the rest of the function if it's temporarily not enabled)

nathanhhughes reviewed (commented):
I'm probably okay merging this in sooner rather than later, but need to figure out versioning before I do that. Both Marcus and Jared were having issues with loading scene graphs created over the summer and I need to figure out (i) what I messed up with versioning and (ii) think through if the new attributes need a version bump

Schmluk commented:
@nathanhhughes 
- Good point! Let me double check and update serialization if need be!
- No rush to merge this in, just thought it might be easier sooner than later.

Schmluk commented:
@nathanhhughes Many thanks for the feedback! These commits should address all points:
- Updated the attribute ordering so the attribute IDs should remain unchanged. Then all serialization should work without extra version bumps as far as I can tell.
  - I wonder if it is worth to somewhere store a (or several) 'old' DSGs and wget & load it as part of the utests to see if it breaks?
- Implemented and tested the missing empty functions, thanks for the catch!
- Now all arguments are optimistic (no confusing mix of optimistic and pessimistic, which was for historical reasons)
